### PR TITLE
[WIP] Data host opt in

### PR DESCRIPTION
**Pull request by melatron**

### CODE REVIEW

**Comment by shelbyd** [comment]
If this depends on #86, please make the base of this PR that branch. It makes it much easier for me to review since I see only what is new in this PR.

File: data_host/chrome_extension/src/pages/Popup/index.tsx [hunk 1]
```diff
@@ -1,23 +1,18 @@
-import React from 'react';
-import { render } from 'react-dom';
-import { Provider } from 'react-redux';
+// import React from 'react';
+// import { render } from 'react-dom';
+// import { Provider } from 'react-redux';
```
Comment (shelbyd) [inline data_host/chrome_extension/src/pages/Popup/index.tsx:3]: Please don't comment code (unless it's for documentation).

File: support/@trustfractal/data-host/src/popup/Environment.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  'development';
+
+const environment: Environment = {
+  FRACTAL_WEBSITE_URL,
+  IS_DEV: NODE_ENV === 'development',
+  PROTOCOL_RPC_ENDPOINT,
+  PROTOCOL_CURRENCY,
+};
+
+export default environment;
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/Environment.ts:25]: We should expose these to clients as configuration with production values being the default. Clients will have their own means of configuring environments, and we shouldn't impose.

File: support/@trustfractal/data-host/src/popup/Environment.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+const FRACTAL_WEBSITE_URL =
+  //   process.env.REACT_APP_FRACTAL_WEBSITE_URL ||
+  'https://staging.sandbox.fractal.id';
+const PROTOCOL_RPC_ENDPOINT =
+  //   process.env.REACT_APP_PROTOCOL_RPC_ENDPOINT ||
+  'wss://nodes.testnet.fractalprotocol.com';
+
+const PROTOCOL_CURRENCY =
+  // process.env.REACT_APP_PROTOCOL_CURRENCY ||
+  'FCL';
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/Environment.ts:12]: Good to remove.

File: support/@trustfractal/data-host/src/popup/Environment.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { Environment } from '@popup/types/Environment';
+
+const FRACTAL_WEBSITE_URL =
+  //   process.env.REACT_APP_FRACTAL_WEBSITE_URL ||
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/Environment.ts:4]: Don't comment out code.

File: support/@trustfractal/data-host/src/popup/components/App.tsx [hunk 5]
```diff
@@ -1,9 +1,12 @@
...
+import Protocol from '@components/Protocol';
+import GlobalStyle from "@styles/GlobalStyle";
 function App() {
 
   return (
+    <>
+    <GlobalStyle />
     <Protocol />
+    </>
   );
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/components/App.tsx:10]: Do we have a formatter configured that handles jsx? I think these are usually indented below the top-level <></>.

File: support/@trustfractal/data-host/src/popup/components/Protocol.tsx [hunk 6]
```diff
@@ -1,10 +1,63 @@
...
+
+        setPageOverride(
+          <SetupSuccess
+            mnemonic={mnemonic}
+            onContinue={() => setPageOverride(null)}
+          />,
+        );
+      } catch (e) {
+        console.log(e);
+        // handleError(e, () => optInWithMnemonic(mnemonic));
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/components/Protocol.tsx:41]: Why remove this?

File: support/@trustfractal/data-host/src/popup/services/DataHost.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
+import { Storage, StorageArray } from '@utils/StorageArray';
+
+export class DataHost {
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/services/DataHost.ts:3]: This is the type of thing we should write in Rust. If we want to include it as TS for now, just to make progress, that's fine. But we should move this into a Rust crate.

File: support/@trustfractal/data-host/package.json [hunk 8]
```diff
@@ -1,25 +1,24 @@
 {
   "name": "@trustfractal/data-host",
```
Comment (shelbyd) [inline support/@trustfractal/data-host/package.json:2]: I'd like to have the Data Host library in /data_host/library or something similar. I think /support should really only be for very minimal focused things that we use to support our own development.

File: support/@trustfractal/data-host/src/popup/services/Factory.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+import { ProtocolService } from '@services/Protocol';
+import { ProtocolOptIn } from '@services/ProtocolOptIn';
+import { StorageService } from '@services/StorageService';
+import { AggregateMultiContext, MultiContext } from '@utils/MultiContext';
+import { ValueCache } from '@utils/ReactHooks';
+
+const types = {
+  FractalId: 'u64',
+  MerkleTree: 'Raw',
+};
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/services/Factory.ts:14]: Maybe later: We should import these from our source of truth, /blockchain/types.json I think.

File: support/@trustfractal/data-host/src/popup/services/Factory.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+let storageService: StorageService;
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/services/Factory.ts:16]: I'm not super happy with having this one factory file. I'd prefer a dependency injection framework, but I didn't want to do the effort of integrating that into the wallet. We can consider changing this.

File: support/@trustfractal/data-host/src/popup/services/FractalAccount.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  constructor(message?: string) {
+    super(message);
+    this.name = 'NotConnectedError';
+  }
+}
+
+const NEXT_TOKENS_KEY = 'fractal-account-connector/will-accept-next-tokens';
+const TOKENS_KEY = 'fractal-account-connector/tokens';
+
+export class FractalAccountConnector extends MultiContext {
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/services/FractalAccount.ts:20]: Most of this class is for our old way of connecting the account. We can remove/change it now or later.

File: support/@trustfractal/data-host/src/popup/services/StorageService/Errors.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
+enum ErrorCode {
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/services/StorageService/Errors.ts:1]: We shouldn't need this file.

File: support/@trustfractal/data-host/src/popup/services/StorageService/index.ts [hunk 13]
```diff
@@ -0,0 +1,74 @@
+import {
+  ERROR_CLEAR,
+  ERROR_GET_ITEM,
+  ERROR_HAS_ITEM,
+  ERROR_REMOVE_ITEM,
+  ERROR_SET_ITEM,
+} from './Errors';
```
Comment (shelbyd) [inline support/@trustfractal/data-host/src/popup/services/StorageService/index.ts:7]: Let's change how this service works to use subclasses of Error like the other services.

**Review by shelbyd — APPROVED** [review]

File: support/@trustfractal/data-host/src/popup/services/FractalAccount.ts (hunk 11 above)
Comment (melatron) [reply support/@trustfractal/data-host/src/popup/services/FractalAccount.ts:20]: By the old way, do you mean that we are currently using another way in the Fractal Wallet? If so, can you send me link to it.

File: support/@trustfractal/data-host/src/popup/services/Factory.ts (hunk 10 above)
Comment (melatron) [reply support/@trustfractal/data-host/src/popup/services/Factory.ts:16]: Do you have on mind a specific dependency injection framework, or should I research?

Comment (shelbyd) [reply support/@trustfractal/data-host/src/popup/services/Factory.ts]: I like Angular's, but I don't know how usable it is outside of Angular projects. Not worth holding up this PR though.

File: support/@trustfractal/data-host/src/popup/services/FractalAccount.ts (hunk 11 above)
Comment (shelbyd) [reply support/@trustfractal/data-host/src/popup/services/FractalAccount.ts]: I mean the way the wallet is using now, but what we're moving away from. You can include this for now, but be sure to remove the unused code once fractal.id registers the identity.

**Pull request merged**